packages/squiggle-lang/src/XYShape.ts

```diff
@@ -640,3 +665,127 @@ export const Analysis = {
+*/
+export const extractSubsetThatSatisfiesThreshold = (
+  shape: XYShape,
+  comparisonType: ComparisonType,
+  threshold: number
+): XYShapeSubset => {
+  const comparisonIsSatisfied = (y: number, _threshold: number): boolean => {
+    switch (comparisonType) {
+      case "greaterThan":
+        return y > _threshold;
+      case "lesserThan":
+        return y < _threshold;
+      case "equals":
+        return Math.abs(y - _threshold) < Number.EPSILON;
+      case "greaterThanOrEqual":
+        return y >= _threshold;
+      case "lessThanOrEqual":
+        return y <= _threshold;
+    }
+  };
+
+  const points: [number, number][] = [];
+  let segments: XYShape[] = [];
+
+  let currentSegment: XYShape | null = null;
+
+  //Assumes linear interpolation
+  function interpolateXAtThreshold(
+    x1: number,
+    x2: number,
+    y1: number,
+    y2: number
+  ) {
+    return x1 + ((threshold - y1) * (x2 - x1)) / (y2 - y1);
+  }
+
+  T.zip(shape).forEach(([x1, y1], i) => {
+    if (i === shape.xs.length - 1) {
+      if (currentSegment) {
+        segments.push(currentSegment);
+      }
+      return;
+    }
+
+    const x2 = shape.xs[i + 1];
+    const y2 = shape.ys[i + 1];
+
+    if (
+      comparisonIsSatisfied(y1, threshold) &&
+      comparisonIsSatisfied(y2, threshold)
+    ) {
+      if (!currentSegment) {
+        currentSegment = { xs: [x1], ys: [y1] };
+      }
+      currentSegment.xs.push(x2);
+      currentSegment.ys.push(y2);
+    } else if (
+      !comparisonIsSatisfied(y1, threshold) &&
+      comparisonIsSatisfied(y2, threshold)
+    ) {
+      const x = interpolateXAtThreshold(x1, x2, y1, y2);
+      if (!currentSegment) {
+        currentSegment = { xs: [x], ys: [threshold] };
+      }
+      currentSegment.xs.push(x2);
+      currentSegment.ys.push(y2);
+    } else if (
+      comparisonIsSatisfied(y1, threshold) &&
+      !comparisonIsSatisfied(y2, threshold)
+    ) {
+      const x = interpolateXAtThreshold(x1, x2, y1, y2);
+      if (currentSegment) {
+        currentSegment.xs.push(x);
+        currentSegment.ys.push(threshold);
+        segments.push(currentSegment);
+        currentSegment = null;
+      }
+    } else if (currentSegment) {
+      segments.push(currentSegment);
+      currentSegment = null;
+    }
+
+    if (
+      comparisonType === "equals" &&
+      y1 !== y2 &&
+      ((y1 < threshold && y2 > threshold) || (y1 > threshold && y2 < threshold))
+    ) {
+      const x = interpolateXAtThreshold(x1, x2, y1, y2);
+      points.push([x, threshold]);
+    }
+  });
+
+  segments = segments.map(T.removeConsecutiveDuplicates);
+  segments.forEach((segment) => {
+    if (T.length(segment) === 1) {
+      points.push([segment.xs[0], segment.ys[0]]);
+    }
+  });
+  segments = segments.filter((segment) => T.length(segment) > 1);
+
+  return {
+    points: sortBy(points, ([x]) => x),
+    segments,
+  };
```

Comment:
General thoughts on this function, without going into details too much:
- there are a few things here that could be optimized (e.g. multiple calls to `comparisonIsSatisfied`), but it's not very important
- I'm kinda sad about it, because by now I did ~3 attempts to rewrite and clean up `XYShape`; it's hard to do incrementally and hard to do in a single day; so the more code this file has, the harder it'll be to do
- but I guess it's worth it and not a blocker for the PR

Reply:
> so the more code this file has, the harder it'll be to do
I agree, but at the same time, I don't expect it to be ``too`` bad to convert this to any other XYShape format, especially with tests + LLMs.

Reply:
Yea, I wasn't very focused on optimization. I don't expect we'll call this too often, and we can optimize if it ever seems like we'll want to, later.